pkg/reconciler/serverlessservice/subsetter.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serverlessservice
+
+import (
+	"hash"
+	"hash/fnv"
+	"sort"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	startSalt = "start-angle-salt"
+	stepSalt  = "step-angle-salt"
+	fromSalt  = "from-salt"
+
+	// universe represents the possible range of angles [0, universe).
+	// We want to have universe divide total range evenly to reduce bias.
+	universe = (1 << 10)
+)
+
+// computeAngle returns a uint64 number swhich represents
+// a hashe built off the given `n` string for consistent selection
+// algorithm.
+func computeHash(n string, h hash.Hash64) uint64 {
+	h.Reset()
+	h.Write([]byte(n))
+	return h.Sum64()
+}
+
+type hashData struct {
+	// The set of all hashes for fast lookup and to name mapping
+	nameLookup map[int]string
+	// Sorted set of hashes for selection algorithm.
+	hashPool []int
+	// start angle
+	start int
+	// step angle
+	step int
+}
+
+func (hd *hashData) fromIndexSet(s sets.Int) sets.String {
+	ret := sets.NewString()
+	for v := range s {
+		ret.Insert(hd.nameForHIndex(v))
+	}
+	return ret
+}
+
+func (hd *hashData) nameForHIndex(hi int) string {
+	return hd.nameLookup[hd.hashPool[hi]]
+}
+
+func buildHashes(from []string, target string) *hashData {
+	hasher := fnv.New64a()
+	hd := &hashData{
+		nameLookup: make(map[int]string, len(from)),
+		hashPool:   make([]int, len(from)),
+		start:      int(computeHash(target+startSalt, hasher) % universe),
+		step:       int(computeHash(target+stepSalt, hasher) % universe),
+	}
+
+	for i, f := range from {
+		// Make unique sets for every target.
+		k := f + target
+		h := int(computeHash(k, hasher))
+		hs := h % universe
+		// Two values slotted to the same bucket.
+		// On average should happen with 1/universe probability.
+		for _, ok := hd.nameLookup[hs]; ok; _, ok = hd.nameLookup[hs] {
```

Comment:
Woah, I needed to read this 3 times until I got it 😂 . Maybe unroll this line to

```golang
_, ok := hd.nameLookup[hs]
for ok {
  ...
  _, ok = hd.nameLookup[hs]
}
```

Reply:
:) Alright